cpu/gd32v/periph/i2c.c

```diff
@@ -0,0 +1,428 @@
+ *               2023 Gunar Schorcht <gunar@schorcht.net>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_gd32v
+ * @ingroup     drivers_periph_i2c
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation
+ *
+ * This driver is a modified copy of the I2C driver for the STM32F1 family.
+ *
+ * @note This implementation only implements the 7-bit addressing polling mode.
+ *
+ * @author      Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Thomas Eichinger <thomas.eichinger@fu-berlin.de>
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Toon Stegen <toon.stegen@altran.com>
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ * @author      Víctor Ariño <victor.arino@triagnosys.com>
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ * @author      Kevin Weiss <kevin.weiss@haw-hamburg.de>
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ *
+ * @}
+ */
+
+#include <assert.h>
+#include <stdint.h>
+#include <errno.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "mutex.h"
+#include "pm_layered.h"
+#include "panic.h"
+
+#include "periph/i2c.h"
+#include "periph/gpio.h"
+#include "periph_conf.h"
+
+/* Some DEBUG statements may cause delays that alter i2c functionality */
+#define ENABLE_DEBUG        0
+#include "debug.h"
+
+#define TICK_TIMEOUT        (0xFFFF)
+
+#define I2C_IRQ_PRIO        (1)
+#define I2C_FLAG_READ       (I2C_READ)
+#define I2C_FLAG_WRITE      (0)
+
+#define ERROR_FLAGS         (I2C_STAT0_AERR_Msk | I2C_STAT0_LOSTARB_Msk | I2C_STAT0_BERR_Msk)
+
+/* static function definitions */
+static void _init(i2c_t dev);
+static void _init_pins(i2c_t dev);
+static void _init_clk(I2C_Type *i2c, uint32_t speed);
+static void _deinit_pins(i2c_t dev);
+
+static int _start(I2C_Type *dev, uint8_t address_byte, uint8_t flags,
+                  size_t length);
+static int _stop(I2C_Type *dev);
+
+static int _is_sr1_mask_set(I2C_Type *i2c, uint32_t mask, uint8_t flags);
+static inline int _wait_for_bus(I2C_Type *i2c);
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[I2C_NUMOF];
+
+void i2c_init(i2c_t dev)
+{
+    assert(dev < I2C_NUMOF);
+
+    mutex_init(&locks[dev]);
+
+    assert(i2c_config[dev].dev != NULL);
+
+    /* Configure pins in idle state as open drain outputs to keep the bus lines
+     * in HIGH state */
+    _deinit_pins(dev);
+
+    periph_clk_en(APB1, i2c_config[dev].rcu_mask);
+
+    _init(dev);
+
+    periph_clk_dis(APB1, i2c_config[dev].rcu_mask);
+}
+
+static void _init_pins(i2c_t dev)
+{
+    /* This is needed in case the remapped pins are used */
+    if (i2c_config[dev].scl_pin == GPIO_PIN(PORT_B, 8) ||
+        i2c_config[dev].sda_pin == GPIO_PIN(PORT_B, 9)) {
```

Comment:
Is this also true for GD32V? IIRC this was about SWD

Reply:
> Is this also true for GD32V?

Yes, I2C0_SCL and I2C0_SDA signals could be remapped to PB8 and PB9 if a board specific configuration is used.

Reply:
> IIRC this was about SWD

IMHO this has nothing to do with SWD. Some of the signals can be remapped to other GPIOs. This is true for GD32V as well as for STM32F1.